src/Config.hs

```diff
@@ -40,50 +47,286 @@ newtype MilliSeconds = MilliSeconds { unMilliSeconds :: Int }
--- This type is internal to the workings of this module. It is used as an
--- intermediate value to get optparse-applicative to play nice with environment
--- variables as used for behavior flags. All flags are off by default.
-data EnvFlags = EnvFlags
-  { efConnectTls :: Bool
-  , efValidateCerts :: Bool
-  , efInheritEnv :: Bool
-  , efUsePath :: Bool
+    [ "Host:           " ++ showSpecifiedString (oVaultHost opts)
+    , "Port:           " ++ showSpecified (oVaultPort opts)
+    , "Addr            " ++ showSpecifiedString (oVaultAddr opts)
+    , "Token:          " ++ maybe "Unspecified" (const "*****") (oVaultToken opts)
+    , "Secret file:    " ++ showSpecifiedString (oSecretFile opts)
+    , "Command:        " ++ showSpecifiedString (oCmd opts)
+    , "Arguments:      " ++ showSpecified (oArgs opts)
+    , "Use TLS:        " ++ showSpecified (oConnectTls opts)
+    , "Validate certs: " ++ showSpecified (oValidateCerts opts)
+    , "Inherit env:    " ++ showSpecified (oInheritEnv opts)
+    , "Base delay:     " ++ showSpecified (unMilliSeconds <$> oRetryBaseDelay opts)
+    , "Retry attempts: " ++ showSpecified (oRetryAttempts opts)
+    , "Log-level:      " ++ showSpecified (oLogLevel opts)
+    , "Use PATH:       " ++ showSpecified (oUsePath opts)
+    ] where
+      showSpecified :: Show a => Maybe a -> String
+      showSpecified (Just x) = show x
+      showSpecified Nothing = "Unspecified"
+      showSpecifiedString = fromMaybe "Unspecified"
+
+-- | Gets the option from a Maybe, ```name``` should only be used for debugging
+-- purposes, as ```isOptionsComplete``` should verify that every property that
+-- does not have a default value, is present in the options.
+getOptionsValue :: (Options Validated Completed -> Maybe c) -> Options Validated Completed -> c
+getOptionsValue f opts=
+    fromMaybe (error "Complete options has an unknown key") (f opts)
+
+-- | The possible errors that can occur in the construction of an Options datatype
+data OptionsError
+  = UnspecifiedValue String -- ^ A value is missing and no default is specified
+  | InvalidAddrFormat String -- ^ The format of the address is invalid
+  | UnknownScheme String -- ^ The scheme of the address is invalid, e.g. ftp://
+  | NonNumericPort String -- ^ The port of the address is not an valid integer
+  | HostPortSchemeAddrMismatch Bool String Int String
+      -- ^ The useTls, host, port and scheme do not match the provided address
+
+
+
+
+instance Show OptionsError where
+  show (UnspecifiedValue s) = "The option " ++ s ++ " is required but not specified"
+  show (InvalidAddrFormat s) = "The address " ++ s ++ " has an invalid format"
+  show (UnknownScheme s)  = "The address " ++ s ++ " has no recognisable scheme, "
+      ++ " expected http:// or https:// at the beginning of the address."
+  show (NonNumericPort s) = "The port " ++ s ++ " is not a valid int value"
+  show (HostPortSchemeAddrMismatch useTLs host port addr) =
```

Comment:
I clarified the error message, it now reads:
```
[ERROR] Confliciting configuration values in cli options
Vault address: "http://localhost:42"
Vault host: Unspecified
Vault port: 1234
Use TLS: Unspecified
Hint: This can happen when you provide a `addr` which conflicts with `port`, `host`, or `[no-]connect-tls` in either the environment variables or the CLI.
```
I changed the hint a bit, because the conflicts can also occur in the environment variables of the .env files.